sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/query/executor/QueryExecutor.java

```diff
@@ -0,0 +1,619 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.stargate.sgv2.docsapi.service.query.executor;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Streams;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.BytesValue;
+import com.google.protobuf.Int32Value;
+import hu.akarnokd.rxjava3.operators.Flowables;
+import io.reactivex.rxjava3.core.Flowable;
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.converters.multi.MultiRx3Converters;
+import io.stargate.bridge.proto.QueryOuterClass;
+import io.stargate.bridge.proto.StargateBridge;
+import io.stargate.sgv2.docsapi.api.common.StargateRequestInfo;
+import io.stargate.sgv2.docsapi.api.common.properties.document.DocumentProperties;
+import io.stargate.sgv2.docsapi.api.common.properties.document.DocumentTableProperties;
+import io.stargate.sgv2.docsapi.service.ExecutionContext;
+import io.stargate.sgv2.docsapi.service.common.model.RowWrapper;
+import io.stargate.sgv2.docsapi.service.query.model.ImmutableRawDocument;
+import io.stargate.sgv2.docsapi.service.query.model.RawDocument;
+import io.stargate.sgv2.docsapi.service.query.model.paging.CombinedPagingState;
+import io.stargate.sgv2.docsapi.service.query.model.paging.PagingStateSupplier;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+/** Executes pre-built document queries, groups document rows and manages document pagination. */
+@ApplicationScoped
+public class QueryExecutor {
```

Comment:
It would really help to have a high-level explanation of this class: why searches can result into multiple CQL queries, the problem of combining paging states, key depth, exponential page sizes, etc.
Maybe with a few concrete examples (like I did in the write query builder classes), if we can find ones that are sufficiently compact yet meaningful.

From an outsider's point of view, the code is very hard to follow.

Reply:
True. Will try to add some explanations..